cucumber-core/src/main/java/io/cucumber/core/runtime/UuidGeneratorServiceLoader.java

```diff
@@ -0,0 +1,115 @@
+        Class<? extends UuidGenerator> objectFactoryClass = options.getUuidGeneratorClass();
+        ClassLoader classLoader = classLoaderSupplier.get();
+        ServiceLoader<UuidGenerator> loader = ServiceLoader.load(UuidGenerator.class, classLoader);
+        if (objectFactoryClass == null) {
+            return loadSingleUuidGeneratorOrDefault(loader);
+        }
+
+        return loadSelectedUuidGenerator(loader, objectFactoryClass);
+    }
+
+    private static UuidGenerator loadSingleUuidGeneratorOrDefault(ServiceLoader<UuidGenerator> loader) {
+        Iterator<UuidGenerator> uuidGenerators = loader.iterator();
+
+        // Find the first UUID generator
+        UuidGenerator uuidGenerator = null;
+        if (uuidGenerators.hasNext()) {
+            uuidGenerator = uuidGenerators.next();
+        }
+        if (uuidGenerator == null) {
+            throw new CucumberException("" +
+                    "Could not find any UUID generator.\n" +
+                    "\n" +
+                    "Cucumber uses SPI to discover UUID generator implementations.\n" +
+                    "This typically happens when using shaded jars. Make sure\n" +
+                    "to merge all SPI definitions in META-INF/services correctly");
+        }
+
+        // look for the DefaultUuidGenerator
+        UuidGenerator extraUuidGenerator = null;
+        while (uuidGenerators.hasNext() && !(uuidGenerator instanceof DefaultUuidGenerator)) {
+            extraUuidGenerator = uuidGenerator;
+            uuidGenerator = uuidGenerators.next();
+        }
+
+        // Check if there are no other non-default object factories
+        if (!(uuidGenerator instanceof DefaultUuidGenerator)) {
+            throw new CucumberException(getMultipleUuidGeneratorLogMessage(uuidGenerator, extraUuidGenerator));
+        }
+
+        return uuidGenerator;
+    }
+
+    private static UuidGenerator loadSelectedUuidGenerator(
+            ServiceLoader<UuidGenerator> loader, Class<? extends UuidGenerator> uuidGeneratorClass
+    ) {
+        for (UuidGenerator uuidGenerator : loader) {
+            if (uuidGeneratorClass.equals(uuidGenerator.getClass())) {
+                return uuidGenerator;
+            }
+        }
+
+        throw new CucumberException("" +
+                "Could not find UUID generator " + uuidGeneratorClass.getName() + ".\n" +
+                "\n" +
+                "Cucumber uses SPI to discover UUID generator implementations.\n" +
+                "Has the class been registered with SPI and is it available on\n" +
+                "the classpath?");
+    }
+
+    private static String getMultipleUuidGeneratorLogMessage(UuidGenerator... uuidGenerators) {
+        String factoryNames = Stream.of(uuidGenerators)
+                .map(Object::getClass)
+                .map(Class::getName)
+                .collect(Collectors.joining(", "));
+
+        return "More than one Cucumber UuidGenerator was found on the classpath\n" +
```

Comment:
This exception needs some consideration. We have multiple generators by default so I don't think it is applicable.

Reply:
I'll continue to work on this part.

Reply:
@mpkorstanje I added some test cases on the `UuidGeneratorServiceLoader` (which revealed one issue regarding the exception message above).

Then I applied the same concepts on `ObjectFactoryServiceLoader`, but kept the tests `@Disabled` because they are not passing: https://github.com/cucumber/cucumber-jvm/pull/2703/files#diff-cbb57a5f00daaf5e93ea9bf64b59ae727113b0af3e5a98b70d06ad5118864766 . Thus, either we have an issue in the tests or in the code.

Could you please confirm me the expected `ObjectFactoryServiceLoader` behavior ?

| Option property `object-factory` | `ObjectFactory` in META-INF/services/... file | Expected behavior                    |
| ----------------------------------- | ------------------------------------------------ | --------------------------------- |
| not specified                               | DefaultObjectFactory                                      | uses DefaultObjectFactory       |
| not specified                               | OtherObjectFactory                                         | uses OtherObjectFactory        |
| not specified                               | none                                                               | exception "could not find any" |
| not specified                               | DefaultObjectFactory + OtherObjectFactory  | exception "more than one"      |
| not specified                               | OtherObjectFactory + DefaultObjectFactory  | exception "more than one"      |

(the two new `@Disabled` test cases cover the last two lines of the table above).

Reply:
I think the current problem is caused by having two "defaults". The easy way around it would be to keep the fast/incrementing UUID generator in your own project.

If we do want to keep both implementations then things get rather complicated because we no longer have a good way to make the current default gracefully step aside. 

So the correct solution would be to 1) non-defaults are prioritized. 2) When only default are available, use the primary default.

But in the interest of simplicity I think having only one "default" implementation is better.

Reply:
I don't think there is a problem with the UUID generator. 

Let's make some use-cases... (I'll keep my question on `ObjectFactory` aside for the moment).
Note: the use-cases are ordered by decreasing estimated popularity.

### Use-case 1: multiple UUID generators from Cucumber

My application uses Cucumber and no other UUID generator is provided. It will be only one  `META-INF/services/io.cucumber.core.eventbus.UuidGenerator` files on the classpath. Thus, when the `UuidGeneratorServiceLoader` runs, the `ServiceLoader.iterator()` will have the following UUID generators (package name not mentioned for simplicity):
- `RandomUuidGenerator` (coming from Cucumber META-INF)
- `IncrementingUuidGenerator` (coming from Cucumber META-INF)

The `uuid-generator` selection parameter is left unspecified.

### Use-case 2: multiple UUID generators from multiple sources

My application uses Cucumber with two other third-party libraries which implement Cucumber UUID generator SPI (respectively `LibraryOneUuidGenerator` and `LibraryTwoUuidGenerator`). 

When the application is packaged in a classical way (i.e. not a fat jar), it will be 3 `META-INF/services/io.cucumber.core.eventbus.UuidGenerator` files on the classpath. 
Thus, when the `UuidGeneratorServiceLoader` runs, the `ServiceLoader.iterator()` will have the following UUID generators (package name not mentioned for simplicity):
- `RandomUuidGenerator` (coming from Cucumber META-INF)
- `IncrementingUuidGenerator` (coming from Cucumber META-INF)
- `LibraryOneUuidGenerator` (coming from LibraryOne META-INF)
- `LibraryTwoUuidGenerator` (coming from LibraryTwo META-INF)

The `uuid-generator` selection parameter is left unspecified.

### Use-case 3: only one UUID generator

My application uses Cucumber and I packaged it as a fat jar with a single one custom `META-INF/services/io.cucumber.core.eventbus.UuidGenerator` files on the classpath. Let's say I specified only one UUID generator in this file, a custom `MyApplicationUuidGenerator`. Thus when the `UuidGeneratorServiceLoader` runs, the `ServiceLoader.iterator()` will have the UUID generators defined in this file:
- `MyApplicationUuidGenerator`

The `uuid-generator` selection parameter is left unspecified.

### Use-case 4: multiple UUID generators without the `RandomUuidGenerator`

My application uses Cucumber and I packaged it as a fat jar with a single one custom `META-INF/services/io.cucumber.core.eventbus.UuidGenerator` files on the classpath. Let's say I specified only two UUID generators in this file, a custom `MyApplicationUuidGenerator` and Cucumber's `IncrementingUuidGenerator`. Thus when the `UuidGeneratorServiceLoader` runs, the `ServiceLoader.iterator()` will have the UUID generators defined in this file:
- `MyApplicationUuidGenerator`
-  `IncrementingUuidGenerator`

The `uuid-generator` selection parameter is left unspecified.

### Use-case 5: no UUID generator

My application uses Cucumber and I packaged it as a fat jar with a single one custom `META-INF/services/io.cucumber.core.eventbus.UuidGenerator` files on the classpath. This file is empty. Thus when the `UuidGeneratorServiceLoader` runs, the `ServiceLoader.iterator()` will have the no UUID generators defined.

The `uuid-generator` selection parameter is left unspecified.

### Expected behavior

| Use-case | UUID generators available on SPI | Expected selected UUID generator |
| -----------| ------------------------------------| ---------------------------------------|
| Use-case 1 | `RandomUuidGenerator` <br> `IncrementingUuidGenerator` |  `RandomUuidGenerator`  |
| Use-case 2 | `RandomUuidGenerator` <br> `IncrementingUuidGenerator` <br> `LibraryOneUuidGenerator` <br> `LibraryTwoUuidGenerator` |  `RandomUuidGenerator`  |
| Use-case 3 | `MyApplicationUuidGenerator` | `MyApplicationUuidGenerator` |
| Use-case 4 | `MyApplicationUuidGenerator` <br> `IncrementingUuidGenerator` | error "more than one UUID generator without selection" |
| Use-case 5 | none | error "no UUID generator found as SPI" |

So the rule is (when no "uuid-generator" selection parameter is provided) :
- when the `RandomUuidGenerator` is present, use it
- when a single `UuidGenerator` is present, use it
- when multiple `UuidGenerator` are present (without `RandomUuidGenerator`), raise an error "more than one UUID generator without selection"
- when no `UuidGenerator` is present, raise an error "no UUID generator found as SPI"

(note that the code must be robust to the UUID generators order in the SPI iterator: it must work no matter in the `RandomUuidGenerator` is the first or the last).

The code in this PR is implemented for the rules and test use-cases above.

Reply:
That's fine for me, I just need to use `IncrementingUuidGenerator` only when specified in the uuid-generator property.

That is (same as you wrote, with changes in **bold**) :

| # | uuid-generator property | Available services | Result |
|-|-|-|-|
 | 1 | undefined |  none | exception, no generators available |
 | 2 | undefined |  RandomUuidGenerator, **IncrementingUuidGenerator** | RandomUuidGenerator used |
 | 3 | RandomUuidGenerator |  RandomUuidGenerator, **IncrementingUuidGenerator** | RandomUuidGenerator used |
 | 4 | undefined |  RandomUuidGenerator, **IncrementingUuidGenerator**, OtherGenerator | OtherGenerator used |
 | 5 | RandomUuidGenerator |  RandomUuidGenerator, **IncrementingUuidGenerator**, OtherGenerator | RandomUuidGenerator used |
 | 6 | undefined |  RandomUuidGenerator, **IncrementingUuidGenerator**, OtherGenerator, YetAnotherGenerator |  exception, cucumber couldn't  decide multiple (non default) generators available |
 | 7 | OtherGenerator |  RandomUuidGenerator, **IncrementingUuidGenerator**, OtherGenerator, YetAnotherGenerator | OtherGenerator used |
 | 8 | **IncrementingUuidGenerator** |  RandomUuidGenerator, **IncrementingUuidGenerator**, OtherGenerator, YetAnotherGenerator | **IncrementingUuidGenerator** used |

I think the `IncrementingUuidGenerator` must be included in Cucumber to be developer-friendly (the developer just need to set a property).

Reply:
If you can make that work, sure.